evm/src/cpu/kernel/asm/bignum.asm

```diff
@@ -0,0 +1,296 @@
+    ADD
+    %decrement
+    // stack: b_end_loc, b_len, a_end_loc, retdest
+    %stack (be, bl, ae) -> (0, 0, ae, be, bl)
+    // stack: carry=0, i=0, a_end_loc, b_end_loc, n=b_len, retdest
+add_loop:
+    // stack: carry, i, a_i_loc, b_i_loc, n, retdest
+    DUP4
+    DUP4
+    // stack: a_i_loc, b_i_loc, carry, i, a_i_loc, b_i_loc, n, retdest
+    %mload_kernel_general
+    SWAP1
+    %mload_kernel_general
+    SWAP1
+    // stack: a[i], b[i], carry, i, a_i_loc, b_i_loc, n, retdest
+    ADD
+    ADD
+    // stack: a[i] + b[i] + carry, i, a_i_loc, b_i_loc, n, retdest
+    %stack (val) -> (val, 256, 256, val)
+    // stack: a[i] + b[i] + carry, 256, 256, a[i] + b[i] + carry, i, a_i_loc, b_i_loc, n, retdest
+    DIV
+    // stack: (a[i] + b[i] + carry) // 256, 256, a[i] + b[i] + carry, i, a_i_loc, b_i_loc, n, retdest
+    SWAP2
+    // stack: a[i] + b[i] + carry, 256, (a[i] + b[i] + carry) // 256, i, a_i_loc, b_i_loc, n, retdest
+    MOD
+    // stack: c[i] = (a[i] + b[i] + carry) % 256, carry_new = (a[i] + b[i] + carry) // 256, i, a_i_loc, b_i_loc, n, retdest
+    DUP4
+    // stack: a_i_loc, c[i], carry_new, i, a_i_loc, b_i_loc, n, retdest
+    %mstore_kernel_general
+    // stack: carry_new, i, a_i_loc, b_i_loc, n, retdest
+    %stack (c, i, a, b) -> (a, b, c, i)
+    // stack: a_i_loc, b_i_loc, carry_new, i, n, retdest
+    %decrement
+    SWAP1
+    %decrement
+    SWAP1
+    %stack (a, b, c, i) -> (c, i, a, b)
+    // stack: carry_new, i, a_i_loc - 1, b_i_loc - 1, n, retdest
+    SWAP1
+    %increment
+    SWAP1
+    // stack: carry_new, i + 1, a_i_loc - 1, b_i_loc - 1, n, retdest
+    DUP5
+    DUP3
+    // stack: i + 1, n, carry_new, i + 1, a_i_loc - 1, b_i_loc - 1, n, retdest
+    EQ
+    %not_bool
+    %jumpi(add_loop)
+add_end:
+    STOP
+    // stack: carry_new, i + 1, a_i_loc - 1, b_i_loc - 1, n, retdest
+    %stack (c, i, a, b, n) -> (c, a)
+    // stack: carry_new, a_i_loc - 1, retdest
+    // If carry = 0, no need to decrement.
+    ISZERO
+    %jumpi(increment_end)
+increment_loop:
+    // stack: cur_loc, retdest
+    DUP1
+    %mload_kernel_general
+    // stack: val, cur_loc, retdest
+    %increment
+    // stack: val+1, cur_loc, retdest
+    %stack (v, l) -> (l, v, l, v)
+    DUP2
+    // stack: cur_loc, val+1, cur_loc, val+1, retdest
+    %mstore_kernel_general
+    // stack: cur_loc, val+1, retdest
+    %decrement
+    // stack: cur_loc - 1, val+1, retdest
+    SWAP1
+    // stack: val+1, cur_loc - 1, retdest
+    %eq_const(256)
+    NOT
+    %jumpi(increment_end)
+    // stack: cur_loc - 1, retdest
+    PUSH 0
+    DUP2
+    // stack: cur_loc - 1, 0, cur_loc - 1, retdest
+    %mstore_kernel_general
+    %jump(increment_loop)
+increment_end:
+    // cur_loc, retdest
+    POP
+    // retdest
+    JUMP
+
+%macro subtract_limb
+    // stack: a_i, b_i, borrow
+    DUP3
+    DUP2
+    SUB
+    // stack: a_i - borrow, a_i, b_i, borrow
+    DUP3
+    // stack: b_i, a_i - borrow, a_i, b_i, borrow
+    GT
```

Comment:
This doesn’t work. Consider: `a_i = 0`, `borrow = 1`. A borrow clearly occurs, but `a_i - borrow == 2**256 - 1` due to underflow, so `borrow_new == 0` for all values of `b_i`.